### PR TITLE
[Docs] Manpage poc

### PR DESCRIPTION
**Pull request by luarss**
@precisionmoon 

TODO
- final docs check

Suggestions 
- add `-no-pager`
- `;#undocumented` for checker off.

### CODE REVIEW

File: src/utl/src/Utl.tcl
Comment (precisionmoon) [inline src/utl/src/Utl.tcl:1]: @luarss, is this the latest version?   User input shouldn't be required to get man path.

**Review by precisionmoon — CHANGES_REQUESTED** [review]
@luarss, I suggest the following chnages:
1) change docs/manpages to docs/man
2) are files under docs/md directory symlinks?
    Can we make the file suffix consistent?   Can everything be *.md suffix like openroad.md instead of .txt?
3) where are contents of docs/md/man3?
4) where will "cat" version of man pages be populated?  Does Util.tcl man command know where to find the "cat" pages?

File: src/OpenRoad.tcl [hunk 1]
```diff
@@ -33,6 +33,9 @@
 ##
 ############################################################################
 
+# source Utl.tcl
+source ../src/utl/src/Utl.tcl
+
```
Comment (luarss) [inline src/OpenRoad.tcl:38]:
```suggestion
```

**Comment by luarss** [comment]
Cho, answers to the following:

- are files under docs/md directory symlinks? **No, currently they are actual files. We will gradually migrate these to the actual READMEs since the format is still not fixed yet**
- Can we make the file suffix consistent? Can everything be *.md suffix like openroad.md instead of .txt? **Yes, will put that on our todo.**
- where are contents of docs/md/man3? **Errors/info/warnings yet to be populated**
- where will "cat" version of man pages be populated? Does Util.tcl man command know where to find the "cat" pages? **Yes they will find the files in `docs/manpages/cat`**

File: src/utl/src/Utl.tcl [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+proc check_valid_man_path { path } {
+  if {[file isdirectory "$path/man1"]} {
+    return true
+  } else {
+    puts "Invalid man path, please retry."
+    return false
+  }
+}
```
Comment (gadfort) [inline src/utl/src/Utl.tcl:37]: these are helper function and should not be exposed to the global namespace.

File: src/utl/src/Utl.tcl [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+proc man { name } {
```
Comment (gadfort) [inline src/utl/src/Utl.tcl:39]: you need to add the `sta::define_cmd_args` for this function.

File: src/utl/src/Utl.tcl [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      }
+
+  if { $man_found == false } {
+    utl::error UTL 100 "Man page for $name is not found."
+  }
+}
+
+proc clear {} {
+    exec clear
+}
```
Comment (gadfort) [inline src/utl/src/Utl.tcl:99]: helper function

File: src/utl/src/Utl.tcl (hunk 2 above)
Comment (luarss) [reply src/utl/src/Utl.tcl]: @gadfort I made some changes, does it look okay?

File: docs/src/test/count_outfiles.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import sys
+
+# Test objective: To count the number of expected files in man2/man3
+
+# This script mimics the build process and checks expected output.
+path = os.path.realpath("md_roff_compat.py")
+docs_dir = os.path.dirname(os.path.dirname(os.path.dirname(path)))
+os.chdir(docs_dir)
+os.system("make clean -s && make preprocess -s")
+os.system("make all -j16 -s")
```
Comment (gadfort) [inline docs/src/test/count_outfiles.py:11]: probably don't want to hardcode the -j16 into the test

File: src/rsz/README.md [hunk 6]
```diff
@@ -92,26 +92,24 @@ estimate_parasitics
 | ----- | ----- |
 | `-placement` or `-global_routing` | Either of these flags must be set. Parasitics are estimated based after placement stage versus after global routing stage. |
 
-### Set Don't Use
+### Set Dont Use
```
Comment (gadfort) [inline src/rsz/README.md:95]: why remove the '?

File: src/utl/src/Utl.tcl [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+sta::define_cmd_args "man" { name\
+                            [-manpath manpath]\
+                            [-no_query]}
+proc man { args } {
+  sta::parse_key_args "man" args \
+    keys {-manpath} flags {-no_query}
+  
+  set name [lindex $args 0]
+  set DEFAULT_MAN_PATH "/usr/local/share/man/cat"
```
Comment (gadfort) [inline src/utl/src/Utl.tcl:12]: shouldn't this be the directory of the `openroad bin/../share/man/cat` as the default?

File: src/rsz/README.md (hunk 6 above)
Comment (luarss) [reply src/rsz/README.md:95]: It is used in regex parsing as manpage filenames, so as to match lowercase Tcl commands `set_dont_use`. I know it's not perfect, but will work on it improving this after the tests are done

File: docs/src/test/count_outfiles.py (hunk 5 above)
Comment (luarss) [reply docs/src/test/count_outfiles.py:11]: I understand, the reason I added that in was to speed up docs compilation which can be highly parallel. Will change to something similar to what we do in `./etc/Build.sh`

File: docs/contrib/DeveloperGuide.md [hunk 8]
```diff
@@ -134,6 +134,13 @@ examples. Use swig to define internal functions to C++ functionality.
 Tcl files can be included by encoding them in CMake into a string that
 is evaluated at run time (See [`Resizer::init()`](../main/src/rsz/src/Resizer.cc)).
 
+:::{Note}
+Please refer to the top-level Tcl formatting [guide](TclFormat.md).
+Our top-level Tcl files, in particular, have to be formatted in this specific
+manner because of the automatic parsing used to convert the READMEs into
+manpages.
```
Comment (rovinski) [inline docs/contrib/DeveloperGuide.md:141]: Is there a way to add a syntax check to the CI? That way it will fail if any changes don't adhere to the required style?

Comment (luarss) [reply docs/contrib/DeveloperGuide.md:141]: We can do that, currently we have 2 syntax checkers in the regression 
1. `man_tcl_check.py` - check for function name existing in 
2. `man_tcl_params.py` - check for function params existing in help, proc and readme. Tests can be waived if `;# checker off` is added to the relevant `sta::define_cmd_args`/ `sta::parse_key_args` lines (example: check_antennas where there are deprecated args that appear in proc but not help)

I am guessing we can move the second one to the CI and remove it from regression test?

Comment (rovinski) [reply docs/contrib/DeveloperGuide.md:141]: As long as there is a check to make sure documentation stays updated it sounds good to me 👍

Comment (luarss) [reply docs/contrib/DeveloperGuide.md:141]: CI results: https://github.com/luarss/OpenROAD/actions/runs/7787611080/job/21235152193

**Comment by luarss** [comment]
@precisionmoon Ready for review. 

For some reason it passes the `ibex_sky130hd` CI in pr-head but not pr-merge, so I am not sure if something is wrong with my branch.

**Comment by precisionmoon** [comment]
@maliberty, can Jack's PR be merged?

**Comment by precisionmoon** [comment]
Jack,

Please do the following and resolve merge conflicts.

1) git fetch origin
2) git merge origin/master
3) manually merge conflicts

Thanks.

Cho

File: .gitignore [hunk 9]
```diff
@@ -30,6 +30,8 @@ docs/main
 messages.txt
 Messages.md
 MessagesFinal.md
+!docs/man
+!docs/man/Makefile
```
Comment (maliberty) [inline .gitignore:34]: Why are these needed?  It seems to fight against docs/.gitignore

File: docs/md/man2/.gitignore
Comment (maliberty) [inline docs/md/man2/.gitignore:1]: needed?

File: docs/md/man3/.gitignore
Comment (maliberty) [inline docs/md/man3/.gitignore:1]: needed?

File: docs/README.md [hunk 10]
```diff
@@ -1,29 +1,51 @@
 # OpenROAD Documentation
 
-This documentation is available at [https://openroad.readthedocs.io/en/latest/](https://openroad.readthedocs.io/en/latest/)
+This repo houses code and raw files to build the on-line documentation (using Sphinx)
```
Comment (maliberty) [inline docs/README.md:3]: The docs/ hierarchy contains the code and raw files....

File: docs/README.md [hunk 11]
```diff
@@ -1,29 +1,51 @@
...
-- `virtualenv`
+## Prerequisites
 
-### Install prerequisites:
+- To install pandoc, refer to this [link](https://github.com/jgm/pandoc/blob/main/INSTALL.md). `apt-get` *should* just work for Ubuntu. 
+- To install sphinx requirements, **create a virtual environment (e.g. conda/virtualenv)** and then run `pip install -r requirements.txt`.
 
-``` shell
-virtualenv .venv
-source .venv/bin/activate
```
Comment (maliberty) [inline docs/README.md:16]: Is there a reason we dropped mention of venv?

File: docs/contrib/ReadmeFormat.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+- Flags/Keys: verify if it is optional or required, then insert the 
+necessary square brackets. Also, keys have to be followed with a specifier
+whereas flags only require the `-flag` itself.
+
+```tcl
+command_abc
+    arg1
+    -key1 key1   
+    [-flag1]
+    [-flagDev]
```
Comment (maliberty) [inline docs/contrib/ReadmeFormat.md:32]: typically positional args come after flags

File: docs/md/man1/openroad.md [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+# NAME
+
+openroad - OpenROAD command-line-interface.
+
+# SYNOPSIS
+
+openroad [-help] [-version] [-no_init] [-exit] [-gui]
+       [-threads count|max] [-log file_name] 
+       [-metrics file_name] [-python]
+       cmd_file
```
Comment (maliberty) [inline docs/md/man1/openroad.md:15]: Missing -no_splash

File: docs/md/man1/openroad.md [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+
+openroad - OpenROAD command-line-interface.
+
+# SYNOPSIS
+
+openroad  [-help]  [-version] [-no_init] [-exit] [-gui]
+       [-threads   count|max]   [-log   file_name]   [-metrics
+       file_name] cmd_file
+
+openroad [-python] cmd_file
```
[the remark this answers is not in the document]
Comment (maliberty) [reply docs/md/man1/openroad.md:17]: I have the same question.

File: docs/md/man1/openroad.md [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+
+**openroad  -threads max -log test.log -metrics test.metrics**: Prints the log and metric 
+respectively to `test.log` and `test.metrics`. Also runs the script with maximum
+ available number of threads.
+
+**openroad -gui**: Launches OpenROAD GUI, which further interactive Tcl commands can be run.
+
+# TCL COMMANDS
+
+These  Tcl-based  commands are callable from within the
```
Comment (maliberty) [inline docs/md/man1/openroad.md:63]: Missing write_cdl

File: docs/src/scripts/extract_utils.py [hunk 16]
```diff
@@ -0,0 +1,144 @@
+import re
```
Comment (maliberty) [inline docs/src/scripts/extract_utils.py:1]: Missing copyright header.  Same for many other new files.
A file scope commenting describing the purpose of the script would be good.

File: docs/src/test/count_outfiles.ok [hunk 17]
```diff
@@ -0,0 +1,170 @@
+File linked successfully.
+ERROR: README.md not found in ../src/cmake
+File linked successfully.
+ERROR: README.md not found in ../src/dbSta
+File linked successfully.
+File linked successfully.
+ERROR: README.md not found in ../src/dpo
```
Comment (maliberty) [inline docs/src/test/count_outfiles.ok:7]: This needs fixing but can be a separate PR.  We should document improve_placement.

File: docs/src/test/count_outfiles.ok [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+stt.md
+Names: 1,        Desc: 1,        Syn: 1,        Options: 1,        Args: 1
+tap.md
+Names: 5,        Desc: 5,        Syn: 5,        Options: 5,        Args: 5
+upf.md
+Names: 11,        Desc: 11,        Syn: 11,        Options: 11,        Args: 11
+utl.md
+Names: 1,        Desc: 1,        Syn: 1,        Options: 1,        Args: 1
+Processing ../src/ant/messages.txt
+Info: 2, Warn: 4, Error: 3
```
Comment (maliberty) [inline docs/src/test/count_outfiles.ok:83]: Is this going to need updating every time we add a new message?

File: docs/src/test/count_outfiles.ok [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+  group_idx: \[
+               ^
+  unexpected eof
+  expecting "%", "\\label", "\\nonumber" or whitespace
+[WARNING] This document format requires a nonempty <title> element.
+  Defaulting to 'stt' as the title.
+  To specify a title, use 'title' in metadata or --metadata title="...".
+[WARNING] This document format requires a nonempty <title> element.
+  Defaulting to 'utl' as the title.
+  To specify a title, use 'title' in metadata or --metadata title="...".
```
Comment (maliberty) [inline docs/src/test/count_outfiles.ok:164]: Can these be fixed?

File: docs/src/test/man_func.ok [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+
+	 • Write abstract Library Exchange Format (.lef) files based  on  cur‐
+	   rent database.
+
+ENVIRONMENT
+FILES
+SEE ALSO
+HISTORY
+BUGS
+				  2023/11/13			   openroad(0)
```
Comment (maliberty) [inline docs/src/test/man_func.ok:205]: Will having a date in the .ok cause it to fail as the date changes?  If not how is it being compared?

File: etc/Build.sh [hunk 21]
```diff
@@ -72,6 +73,9 @@ while [ "$#" -gt 0 ]; do
         -no-gui)
             cmakeOptions+=" -DBUILD_GUI=OFF"
             ;;
+        -build-man)
+            cmakeOptions+=" -DBUILD_MAN=ON"
+            ;;
```
Comment (maliberty) [inline etc/Build.sh:78]: How long does this take?  I'm wondering if it should be the default.

File: src/ant/README.md [hunk 22]
```diff
@@ -62,16 +64,16 @@ about this tool.
 ## Algorithm
 
 | <img src="./doc/images/example_ant.png" width=400px> | <img src="./doc/images/step1.png" width=400px> |
-|:--:|:--:|
-| Antenna Checker Algorithm: WireGraph Example | Step 1: (a) Start from the root node (ITerm) using upper Via to find a node for a new wire. (b) Save the ITerm area for cumulative gate/diffusion area. |
+| --- | --- |
+| <b>Antenna Checker Algorithm</b>: WireGraph Example | <b>Step 1</b>: (a) Start from the root node (ITerm) using upper Via to find a node for a new wire. (b) Save the ITerm area for cumulative gate/diffusion area. |
```
Comment (maliberty) [inline src/ant/README.md:68]: Isn't `**` more idiomatic for bold in .md ?  Why are we switching to html markup (<b>)?

File: src/ant/src/AntennaChecker.tcl [hunk 23]
```diff
@@ -30,12 +30,12 @@
...
 
 sta::define_cmd_args "check_antennas" { [-verbose]\
-                                          [-net net]}
+                                          [-net net]};# checker off
 
 proc check_antennas { args } {
   sta::parse_key_args "check_antennas" args \
     keys {-report_file -net} \
-    flags {-verbose -report_violating_nets}
+    flags {-verbose -report_violating_nets};# checker off
```
Comment (maliberty) [inline src/ant/src/AntennaChecker.tcl:38]: Why is the checker off here?

File: src/cts/src/TritonCTS.tcl [hunk 24]
```diff
@@ -85,12 +85,12 @@ sta::define_cmd_args "clock_tree_synthesis" {[-wire_unit unit]
                                              [-num_static_layers] \
                                              [-sink_clustering_buffer] \
                                              [-obstruction_aware] \
-					     [-apply_ndr] \
+					                                   [-apply_ndr] \
```
Comment (maliberty) [inline src/cts/src/TritonCTS.tcl:88]: Whitespace error

File: src/cts/src/TritonCTS.tcl [hunk 25]
```diff
@@ -85,12 +85,12 @@ sta::define_cmd_args "clock_tree_synthesis" {[-wire_unit unit]
...
                                              [-insertion_delay] \
                                              [-sink_buffer_max_cap_derate] \
                                              [-use_dummy_load] \
                                              [-delay_buffer_derate] \
-                                            }
+                                            };# checker off
```
Comment (maliberty) [inline src/cts/src/TritonCTS.tcl:93]: why off?

File: src/gpl/README.md [hunk 26]
```diff
@@ -114,42 +123,45 @@ global_placement
...
 get_global_placement_uniform_density -pad_left -pad_right 
 ```
 
-### Cluster Flops
-
-Cluster single bit flops into multi-bit flops.
+Example scripts demonstrating how to run `gpl` on a sample design on `core01` as follows:
 
+```shell
+./test/core01.tcl
```
Comment (maliberty) [inline src/gpl/README.md:163]: should this be `openroad ./test/core01.tcl`?

File: src/odb/README.md [hunk 27]
```diff
@@ -89,15 +78,327 @@ database to have exactly the same layout across save/restores.
...
+| ----- | ----- |
+| `file_name` | Macro placement file name. |
+
+
+
+
 ## Example scripts
 
+After building successfully, run OpenDB Tcl shell using
+`../../build/src/odb/src/swig/tcl/odbtcl`. An example usage:
```
Comment (maliberty) [inline src/odb/README.md:381]: We dropped user support for odbtcl quite a while ago.  Such commands should just be run in OR.   The binary is still there just for testing; it is no longer installed outside of the build/ hierarchy.

File: src/pad/src/pad.tcl [hunk 28]
```diff
@@ -138,13 +138,14 @@ sta::define_cmd_args "make_io_sites" {-horizontal_site site \
...
-                                      [-ring_index index]}
+                                      [-ring_index index]
+};# checker off
 
 proc make_io_sites {args} {
   sta::parse_key_args "make_io_sites" args \
     keys {-horizontal_site -vertical_site -corner_site -offset -rotation \
       -rotation_horizontal -rotation_vertical -rotation_corner -ring_index} \
-    flags {}
+    flags {}; # checker off
```
Comment (maliberty) [inline src/pad/src/pad.tcl:148]: why off?

File: src/pad/src/pad.tcl [hunk 29]
```diff
@@ -353,7 +354,7 @@ sta::define_hidden_cmd_args "make_fake_io_site" {-name name \
 proc make_fake_io_site {args} {
   sta::parse_key_args "make_fake_io_site" args \
     keys {-name -width -height} \
-    flags {}
+    flags {};# checker off
```
Comment (maliberty) [inline src/pad/src/pad.tcl:357]: This is a public command despite its odd name

File: src/pdn/src/pdn.tcl [hunk 30]
```diff
@@ -854,7 +858,7 @@ proc define_pdn_grid { args } {
   sta::parse_key_args "define_pdn_grid" args \
     keys {-name -voltage_domains -pins -starts_with -obstructions -power_switch_cell \
       -power_control -power_control_network} \
-    flags {}
+    flags {};# checker off
```
Comment (maliberty) [inline src/pdn/src/pdn.tcl:861]: public command

File: src/utl/README.md [hunk 31]
```diff
@@ -0,0 +1,63 @@
+# Utilities
+
+The utility module houses of the manpage command in OpenROAD,
```
Comment (maliberty) [inline src/utl/README.md:3]: I would reword this to "The utility module contains the man command.".  However it feels redundant given that the next thing is the man command itself.

File: src/utl/README.md [hunk 32]
```diff
@@ -0,0 +1,63 @@
...
+- Parameters in square brackets `[-param param]` are optional.
+- Parameters without square brackets `-param2 param2` are required.
+```
+
+### Man
+
+This can be used for a range of commands in different levels as follows:
+- `man1`: Top-level openroad command (e.g. `man openroad`)
+- `man2`: Individual module commands (e.g. `man clock_tree_synthesis`)
+- `man3`: Info, error, warning messages (e.g. `man CTS-0001`)
```
Comment (maliberty) [inline src/utl/README.md:18]: Confusing as it sounds like there are man1/man2/man3 commands rather than three usages of the one man command.

**Review by maliberty — CHANGES_REQUESTED** [review]
I see a lot of nice improvements to the documentation along with the new command.

File: src/pad/src/pad.tcl (hunk 29 above)
Comment (luarss) [reply src/pad/src/pad.tcl:357]: May I ask what does "public command" mean? And also why does it being public mean we are ignoring checks for it?

File: docs/src/test/man_func.ok (hunk 20 above)
Comment (luarss) [reply docs/src/test/man_func.ok:205]: It will not. The date in the source file (openroad.md) is static

File: etc/Build.sh (hunk 21 above)
Comment (luarss) [reply etc/Build.sh:78]: About 2-3 mins with `-j16` if I recall correctly.

File: src/pad/src/pad.tcl (hunk 29 above)
Comment (maliberty) [reply src/pad/src/pad.tcl:357]: By public I mean this is a command we expect some users to use (unlike say a debug command).  I'm wondering why we are ignoring checks for such a command.

File: src/ant/src/AntennaChecker.tcl (hunk 23 above)
Comment (luarss) [reply src/ant/src/AntennaChecker.tcl:38]: Checker is off because of the deprecated arguments in `check_antennas`. As such some flags/keys will exist in proc but not help.

File: src/cts/src/TritonCTS.tcl (hunk 25 above)
Comment (luarss) [reply src/cts/src/TritonCTS.tcl:93]: image
Obsolete flag

File: src/odb/README.md (hunk 27 above)
Comment (luarss) [reply src/odb/README.md:381]: Do you recommend we remove this line?

File: src/pad/src/pad.tcl (hunk 28 above)
Comment (luarss) [reply src/pad/src/pad.tcl:148]: image

Deprecated key

File: src/odb/README.md (hunk 27 above)
Comment (maliberty) [reply src/odb/README.md:381]: I would remove any mention of odbtcl

File: docs/src/test/count_outfiles.ok (hunk 18 above)
Comment (luarss) [reply docs/src/test/count_outfiles.ok:83]: Yes.

Comment (maliberty) [reply docs/src/test/count_outfiles.ok:83]: That is too much overhead.  Can you setup a more local test?

File: src/pad/src/pad.tcl (hunk 29 above)
Comment (luarss) [reply src/pad/src/pad.tcl:357]: Oh I see. We are ignoring checks because it does not appear in `help`. We are only making sure the counts in `help`, `proc` and `README` match

Comment (maliberty) [reply src/pad/src/pad.tcl:357]: @gadfort is there a reason for make_fake_io_site not to be in help?  Its in the README

Comment (gadfort) [reply src/pad/src/pad.tcl:357]: I don't see why it should be excluded.

Comment (maliberty) [reply src/pad/src/pad.tcl:357]: https://github.com/The-OpenROAD-Project/OpenROAD/blob/db6d999c3804c3c433e85bb7a5270826b1579f34/src/pad/src/pad.tcl#L349 shows sta::define_hidden_cmd_args is used

Comment (gadfort) [reply src/pad/src/pad.tcl:357]: Ahh, that is an oversight on my part. It should be define_cmd_args instead

Comment (maliberty) [reply src/pad/src/pad.tcl:357]: @luarss please fix this in this PR to avoid any merge conflicts.

File: docs/README.md (hunk 11 above)
Comment (luarss) [reply docs/README.md:16]: There can be many ways the user might install the python requirements and activate the environment, I can add it back in if its preferred

File: docs/md/man1/openroad.md (hunk 13 above)
Comment (luarss) [reply docs/md/man1/openroad.md:15]: `-no_splash` is currently not showing when we do openroad -help. Is this intended?

File: src/ant/README.md (hunk 22 above)
Comment (luarss) [reply src/ant/README.md:68]: Changed to `**`.

File: src/pdn/src/pdn.tcl (hunk 30 above)
Comment (luarss) [reply src/pdn/src/pdn.tcl:861]: For this, I turned off checking because its flags are spread over different sub-functions: `define_pdn_grid_macro`, `define_pdn_grid_existing` and `define_pdn_grid`. I think this case is pretty rare, we can fix in a future PR maybe?

File: src/utl/README.md (hunk 32 above)
Comment (luarss) [reply src/utl/README.md:18]: Changed to "Level 1/2/3".

File: docs/md/man1/openroad.md (hunk 13 above)
Comment (maliberty) [reply docs/md/man1/openroad.md:15]: no, please fix it

File: docs/src/test/count_outfiles.ok (hunk 18 above)
Comment (luarss) [reply docs/src/test/count_outfiles.ok:83]: Just to clarify, there is also the `man-to-tcl` check that needs to be updated when Tcl code is updated (when they add a new flag/key for eg.)

I will add in 2 tests for each module, correct?

Comment (luarss) [reply docs/src/test/count_outfiles.ok:83]: @maliberty I have added 2 new local tests to `src/ant`, and modified the `docs/src/test` to be depending on static files instead. Could you have a look?
`ant_man_tcl_check` checks for man,proc, readme counts
`ant_readme_msgs_check` checks for readme counts and messages count.

**Comment by luarss** [comment]
@maliberty I have added new changes to local tests above. Would love your feedback. Thanks!

File: docs/src/test/count_outfiles.ok [hunk 33]
```diff
@@ -100,25 +100,25 @@ Info: 5, Warn: 1, Error: 4
...
 Processing ../src/ifp/messages.txt
 Info: 4, Warn: 7, Error: 17
 Processing ../src/mpl/messages.txt
-Info: 12, Warn: 7, Error: 9
+Info: 12, Warn: 7, Error: 8
 Processing ../src/mpl2/messages.txt
 Info: 2, Warn: 6, Error: 21
 Processing ../src/odb/messages.txt
-Info: 64, Warn: 199, Error: 136
+Info: 64, Warn: 199, Error: 137
```
Comment (maliberty) [inline docs/src/test/count_outfiles.ok:113]: It looks like we are still testing against the number of messages across the whole program.  I'm not clear how this is more local.

File: docs/src/test/translator.txt [hunk 34]
```diff
@@ -0,0 +1,94 @@
+GPL 0002 placerBase.cpp:778        DBU: {} INFO https://github.com/The-OpenROAD-Project/OpenROAD/tree/master/workspace/src/gpl/src/placerBase.cpp#L778
+GPL 0003 placerBase.cpp:806        SiteSize: {} {} INFO https://github.com/The-OpenROAD-Project/OpenROAD/tree/master/workspace/src/gpl/src/placerBase.cpp#L806
```
Comment (maliberty) [inline docs/src/test/translator.txt:2]: Will this test fail if line numbers change?

**Comment by maliberty** [comment]
You have a merge conflict to address in src/mpl2/src/mpl.tcl

File: docs/src/test/translator.txt (hunk 34 above)
Comment (luarss) [reply docs/src/test/translator.txt:2]: Apologies. This `count_outfiles` test will be removed. I added a prototype 2 new local tests to src/ant, and modified the docs/src/test to be depending on static files instead. Could you have a look?
ant_man_tcl_check checks for man,proc, readme counts
ant_readme_msgs_check checks for readme counts and messages count.

File: docs/src/test/count_outfiles.ok (hunk 33 above)
Comment (luarss) [reply docs/src/test/count_outfiles.ok:113]: Apologies. This `count_outfiles` test will be removed. I added a prototype 2 new local tests to src/ant, and modified the docs/src/test to be depending on static files instead. Could you have a look?
ant_man_tcl_check checks for man,proc, readme counts
ant_readme_msgs_check checks for readme counts and messages count.

File: docs/src/test/translator.txt (hunk 34 above)
Comment (luarss) [reply docs/src/test/translator.txt:2]: This is now static and independent from the original gpl `messages.txt`, so no.

File: src/cts/src/TritonCTS.tcl [hunk 35]
```diff
@@ -103,7 +103,7 @@ proc clock_tree_synthesis { args } {
           -sink_clustering_levels -tree_buf \
           -sink_buffer_max_cap_derate -delay_buffer_derate} \
     flags {-post_cts_disable -sink_clustering_enable -balance_levels \
-           -obstruction_aware -apply_ndr -insertion_delay -use_dummy_load
+           -obstruction_aware -apply_ndr -insertion_delay -no_insertion_delay -use_dummy_load
```
Comment (maliberty) [inline src/cts/src/TritonCTS.tcl:106]: Something is off here ` -insertion_delay -no_insertion_delay` only one of these should exist and the other should be the default.  @precisionmoon please clarify

Comment (precisionmoon) [reply src/cts/src/TritonCTS.tcl:106]: -insertion_delay is a no op option.  It is there so existing scripts don't error out.

Comment (maliberty) [reply src/cts/src/TritonCTS.tcl:106]: Then it should be removed from the documentation and marked deprecated.  Also in the README.md -no_insertion_delay is in the options list but not the command summary which -insertion_delay is the opposite.

Comment (precisionmoon) [reply src/cts/src/TritonCTS.tcl:106]: Fixed in https://github.com/The-OpenROAD-Project/OpenROAD/pull/4788

**Comment by maliberty** [comment]
@luarss please resolve the conflict so we can get this merged.

**Review by maliberty — APPROVED** [review]
